tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,482 @@
+    delete_sender_socket(sender_socket);
+}
+
+void
+connect_redirect_tests_common(_In_ const struct bpf_object* object, bool dual_stack, _In_ test_addresses_t& addresses)
+{
+    // First cateogory is authorize tests.
+    const char* protocol_string = (_globals.protocol == IPPROTO_TCP) ? "TCP" : "UDP";
+    const char* family_string = (_globals.family == AF_INET) ? "IPv4" : "IPv6";
+    const char* dual_stack_string = dual_stack ? "Dual Stack" : "No Dual Stack";
+
+    // Loopback address.
+    printf("AUTH: Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.loopback_address);
+
+    // Remote address.
+    printf("AUTH: Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.remote_address);
+
+    // Local non-loopback address.
+    printf("AUTH: Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.local_address);
+
+    // Second category is connection redirection tests.
+
+    // Remote -> Remote
+    printf("REDIRECT: Remote -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.remote_address, dual_stack);
+
+    // Remote -> Loopback
+    printf("REDIRECT: Remote -> Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.loopback_address, dual_stack);
+
+    // Remote -> Local non-loopback
+    printf("REDIRECT: Remote -> Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.local_address, dual_stack);
+
+    // Loopback -> Remote
+    printf("REDIRECT: Loopback -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.loopback_address, addresses.remote_address, dual_stack);
+
+    // Loopback -> Local non-loopback
+    printf("REDIRECT: Loopback -> Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.loopback_address, addresses.local_address, dual_stack);
+
+    // Local non-loopback -> Loopback
+    printf("REDIRECT: Local -> Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.local_address, addresses.loopback_address, dual_stack);
+
+    // Local non-loopback -> Remote
+    printf("REDIRECT: Local -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.local_address, addresses.remote_address, dual_stack);
+}
+
+TEST_CASE("connect_redirect_tcp_v4", "[connect_redirect_tests]")
```

Comment:
Have a common function that takes IPPROTO and address family as parameters and use MACROs to define the various test cases instead of copy pasting same code over and over.

Reply:
Done.